Add unstable `rustc-check-cfg` build script output

Urgau opened the pull request:
This PR adds a new build script output as unstable behind `-Zcheck-cfg=output`: `rustc-check-cfg`.

### What does this PR try to resolve?

This PR add a way to add to use the unstable `--check-cfg` command line option of `rustc` and `rustdoc`.

It was discover in [Bump bootstrap compiler to 1.61.0 beta](https://github.com/rust-lang/rust/pull/95678#discussion_r842803445) that `rustc_llvm` sets some custom `cfg` from a build script and because `--check-cfg=values()` is globally enable in the Rust codebase that cause the compilation to fail. For now no values are checked in stage 0 for the entire codebase which is a shame and should be fixed with the addition of this feature.

### How should we test and review this PR?

Commits are separated in: implementation, tests and doc.

Testing should simply be done by adding a valid `cargo:rustc-check-cfg` in a build script.
Watch the added tests or doc to have an example.

### Additional information

This PR is also the logical next step after `-Zcheck-cfg-features`.

rust-highfive commented:
r? @ehuss

(rust-highfive has picked a reviewer for you, use r? to override)

bors commented:
:umbrella: The latest upstream changes (presumably #10486) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #10609) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #10611) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

ehuss reviewed (commented):
This doesn't seem to pass the `--check-cfg` flag to rustdoc, was that intentional?

Urgau commented on src/doc/src/reference/build-scripts.md:
```diff
@@ -245,6 +247,21 @@ identifier, the value should be a string.
 [option-cfg]: ../../rustc/command-line-arguments.md#option-cfg
 
+<a id="rustc-check-cfg"></a>
+#### `cargo:rustc-check-cfg=CHECK_CFG`
```
Sorry, fixed.

Urgau commented on tests/testsuite/build_script.rs:
```diff
@@ -4837,3 +4837,151 @@ for more information about build script outputs.
+            r#"fn main() { println!("cargo:rustc-check-cfg=names(foo)"); }"#,
+        )
+        .build();
+    p.cargo("build -v -Zextra-check-cfg")
+        .masquerade_as_nightly_cargo()
+        .run();
```
I guess it's checking the absence of error without checking the presence of something. To be honest I just copied this test from `cfg`: https://github.com/rust-lang/cargo/blob/cfb89695d167b601f52b82c19ef5975f1f4b2a98/tests/testsuite/build_script.rs#L2317-L2337

Urgau commented:
> This doesn't seem to pass the `--check-cfg` flag to rustdoc, was that intentional?

No, not at all. Fixed and added a test for it.

I also fixed or responded to the review comments.
The PR should be ready for another review.

Urgau commented:
I've updated the PR to use the improvements of https://github.com/rust-lang/cargo/pull/10566.
The tests are now in the `check_cfg` module, they are now also activated on Windows and this addition is now gated under `-Zcheck-cfg=output` (ie. the output of the build script, not sure that a great name).

@rustbot ready

ehuss commented on tests/testsuite/check_cfg.rs:
```diff
@@ -458,3 +458,173 @@ fn features_doc() {
+        )
+        .build();
+
+    p.cargo("build -v -Zcheck-cfg=output")
+        .masquerade_as_nightly_cargo()
+        .run();
```
Can this add a check that it is passing the flag appropriately?

I think that would be something like `.with_stderr_contains(x!("rustc" => "names" of "foo"))` with what you have here?

ehuss commented on src/doc/src/reference/unstable.md:
```diff
@@ -1211,6 +1212,28 @@ cargo check -Z unstable-options -Z check-cfg=values
+`--check-cfg` flag to the compiler. This may be used for compile-time
+detection of unexpected conditional compilation name and/or values.
+
+This can only be used is combination with `-Zcheck-cfg=output` otherwise it's just a nop.
```
```suggestion
This can only be used in combination with `-Zcheck-cfg=output` otherwise it is ignored with a warning.
```

Urgau commented on src/cargo/core/compiler/mod.rs:
```diff
@@ -718,6 +724,12 @@ fn rustdoc(cx: &mut Context<'_, '_>, unit: &Unit) -> CargoResult<Work> {
                 for cfg in output.cfgs.iter() {
                     rustdoc.arg("--cfg").arg(cfg);
                 }
+                if !output.check_cfgs.is_empty() {
+                    rustdoc.arg("-Zunstable-options");
+                    for check_cfg in &output.check_cfgs {
+                        rustdoc.arg("--check-cfg").arg(check_cfg);
+                    }
+                }
```
@ehuss Sound good. Do you want to do it in this PR or as a follow-up ? (I feel like it doesn't really make sense to do it in this PR, but your call)

ehuss replied:
Here is good because it is adding the `--check-cfg` arg.

Urgau replied:
Done (expect for doc-tests because of it's specificity that make it unable to use the new function).

ehuss commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit f2a1e4363f0ea4f600193fec00cf1f505767547b has been approved by `ehuss`

<!-- @bors r=ehuss f2a1e4363f0ea4f600193fec00cf1f505767547b -->
<!-- homu: {"type":"Approved","sha":"f2a1e4363f0ea4f600193fec00cf1f505767547b","approver":"ehuss"} -->

bors commented:
:hourglass: Testing commit f2a1e4363f0ea4f600193fec00cf1f505767547b with merge c8c6e339967617bca15195748b0a72a2a73e486d...
<!-- homu: {"type":"BuildStarted","head_sha":"f2a1e4363f0ea4f600193fec00cf1f505767547b","merge_sha":"c8c6e339967617bca15195748b0a72a2a73e486d"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/runs/6517797716?check_suite_focus=true)
Approved by: ehuss
Pushing c8c6e339967617bca15195748b0a72a2a73e486d to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"ehuss","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/runs/6517797716?check_suite_focus=true"},"merge_sha":"c8c6e339967617bca15195748b0a72a2a73e486d"} -->
